### ISSUE TITLE
Bring back ndistinct

### ISSUE THREAD

**Issue by alastairmatheson** [PHSKC-APDE/rads#162]
The ndistinct metric was removed in v1.0.0. I had been using the option so am now this: 
![image](https://user-images.githubusercontent.com/19675113/147602555-4f20ae4a-942b-4792-bff3-a1631f75e439.png)

Can you please add it back in when you get a chance?

Thanks!

**Comment by dcaseykc** [comment]
@alastairmatheson: Can you provide some details about your use case? Danny and I decided to scrap it in #118 because it was not clear to us what it is supposed to capture. For factors, do you capture the number of levels or the number of unique values within a level (e.g. 1). For numeric categories (e.g. age), I feel like the number could be high enough to make it useless.

**Comment by alastairmatheson** [comment]
I was using it in the OSPI data to count the number of distinct student IDs (`what = "Student_key"`). For now I've replaced it with `summarize = n_distinct(Student_key)` from dplyr but this simple fix only works on single years of data. It would be a bit more effort to replicate the window function from calc.

For factor or character values, I would expect to see the number of levels/number of distinct values. That's analogous to the n_distinct function in dplyr. You could even do the same thing for numeric values (it can be useful to know how many different ages were reported for example).

**Comment by dcaseykc** [comment]
When `what` is a factor, `calc` treats it as an additional by value (otherwise things like `mean` would be nonsensical). In this case, what do you think the return should be? Still the number of levels? If so, the number of levels present in the subset or the number of possible levels?

@dcolombara : What do you think? Should we add ndistinct back?

**Comment by dcolombara** [comment]
@dcaseykc: Ideally, I think the return should be the number of levels in the entire  dataset. If a given level doesn't exist in the subsetted data, then it can simply return that row with zero. Hopefully this will force people to reckon that their filtering/subsetting has eliminated entire categories from the what var. At the same time, just presenting the number of levels in the subsetted data is fine and reasonable. 

If people like it, I'm fine adding ndistinct back in.

**Comment by alastairmatheson** [comment]
If `what` is a factor, what is a mean calculated against?
I still am inclined to go with returning the number of levels in the data that is fed into calc (be it the entire dataset or a subset)

**Comment by dcaseykc** [comment]
For a factor, the mean is the relative fraction of that level. e.g. .6 A, .4 B

**Comment by dcaseykc** [comment]
@dcolombara @alastairmatheson : Does this match your expectations?
``` r
library('rads')
library('survey')
library('data.table')
library('dtsurvey')
library('testthat')
set.seed(98104)


data(api) #from the survey package
sur = dtsurvey(apiclus1, psu = 'dnum', weight = 'pw')
#normal vars
r1 = calc(sur, what = 'awards', metrics = c('ndistinct'), proportion = T, ci = .95)
r2 = calc(sur, what = 'awards', where = awards == 'Yes', metrics = c('ndistinct'), proportion = T, ci = .95)
r3 = calc(sur, what = 'enroll', metrics = c('ndistinct'), by = 'stype', proportion = F, ci = .95)

expect_equal(r1[, unique(ndistinct)], sur[, length(unique(awards))])
expect_equal(r2[, unique(ndistinct)], 1)
expect_equal(r3[, ndistinct], sur[, length(unique(enroll)), stype][, V1])
```

**Comment by alastairmatheson** [comment]
Yes, looks great. Thanks.

**Comment by dcaseykc** [comment]
Fixed by #167
